File: app/Services/StoryService.swift
import Foundation

/// Service for loading and managing stories from the backend API
class StoryService {
    private var cachedStories: [Story] = []
    private let baseURL = "http://localhost:8000"

    /// Load all stories from the backend API
    func loadAllStories() async -> [Story] {
        if !cachedStories.isEmpty {
            return cachedStories
        }

        guard let url = URL(string: "\(baseURL)/api/stories") else {
            print("StoryService: Invalid URL")
            return []
        }

        do {
            let (data, response) = try await URLSession.shared.data(from: url)

            guard let httpResponse = response as? HTTPURLResponse,
                  httpResponse.statusCode == 200 else {
                print("StoryService: Bad response from server")
                return []
            }

            let decoder = JSONDecoder()
            let storyList = try decoder.decode([StoryListItem].self, from: data)

            // Fetch full story details for each story
            var stories: [Story] = []
            for item in storyList {
                if let story = await loadStory(byId: item.id) {
                    stories.append(story)
                }
            }

            // Sort stories by JLPT level (N5 first) then by title
            stories.sort { story1, story2 in
                if story1.metadata.jlptLevel != story2.metadata.jlptLevel {
                    return story1.metadata.jlptLevel.sortOrder < story2.metadata.jlptLevel.sortOrder
                }
                return story1.metadata.title < story2.metadata.title
            }

            cachedStories = stories
            return stories
        } catch {
            print("StoryService: Error loading stories: \(error)")
            return []
        }
    }

    /// Load stories for a specific JLPT level
    func loadStories(for level: JLPTLevel) async -> [Story] {
        let allStories = await loadAllStories()
        return allStories.filter { $0.metadata.jlptLevel == level }
    }

    /// Load a story by its ID from the backend
    func loadStory(byId id: String) async -> Story? {
        guard let url = URL(string: "\(baseURL)/api/stories/\(id)") else {
            print("StoryService: Invalid URL for story \(id)")
            return nil
        }

        do {
            let (data, response) = try await URLSession.shared.data(from: url)

            guard let httpResponse = response as? HTTPURLResponse,
                  httpResponse.statusCode == 200 else {
                print("StoryService: Bad response for story \(id)")
                return nil
            }

            let decoder = JSONDecoder()
            let story = try decoder.decode(Story.self, from: data)
            return story
        } catch {
            print("StoryService: Error loading story \(id): \(error)")
            return nil
        }
    }

    /// Get stories filtered by level
    func stories(for level: JLPTLevel) async -> [Story] {
        let allStories = await loadAllStories()
        return allStories.filter { $0.metadata.jlptLevel == level }
    }

    /// Get stories grouped by JLPT level
    func storiesGroupedByLevel() async -> [JLPTLevel: [Story]] {
        let allStories = await loadAllStories()
        return Dictionary(grouping: allStories) { $0.metadata.jlptLevel }
    }

    /// Clear cached stories (useful for refreshing)
    func clearCache() {
        cachedStories = []
    }
}

/// Summary view of a story for listing (matches backend response)
struct StoryListItem: Codable {
    let id: String
    let title: String
    let titleJapanese: String?
    let jlptLevel: String
    let wordCount: Int
    let genre: String
    let summary: String
    let coverImageURL: String?
    let chapterCount: Int
}

File: backend/app/services/story_service.py
"""Service for loading and managing stories"""
import json
from pathlib import Path
from typing import List, Optional
from app.models.story import Story, StoryListItem


class StoryService:
    """Service for loading stories from JSON files"""

    def __init__(self, stories_path: Path):
        self.stories_path = stories_path
        self._stories_cache: dict[str, Story] = {}
        self._load_stories()

    def _load_stories(self) -> None:
        """Load all stories from JSON files into cache"""
        if not self.stories_path.exists():
            return

        for json_file in self.stories_path.glob("*.json"):
            try:
                with open(json_file, "r", encoding="utf-8") as f:
                    data = json.load(f)
                    story = Story(**data)
                    self._stories_cache[story.id] = story
            except Exception as e:
                print(f"Error loading story from {json_file}: {e}")

    def get_all_stories(self) -> List[StoryListItem]:
        """Get list of all stories (summary view)"""
        return [
            StoryListItem(
                id=story.id,
                title=story.metadata.title,
                titleJapanese=story.metadata.titleJapanese,
                jlptLevel=story.metadata.jlptLevel,
                wordCount=story.metadata.wordCount,
                genre=story.metadata.genre,
                summary=story.metadata.summary,
                coverImageURL=story.metadata.coverImageURL,
                chapterCount=len(story.chapters),
                isPremium=story.metadata.isPremium
            )
            for story in self._stories_cache.values()
        ]

    def get_story_by_id(self, story_id: str) -> Optional[Story]:
        """Get a specific story by ID"""
        return self._stories_cache.get(story_id)

    def get_stories_by_level(self, jlpt_level: str) -> List[StoryListItem]:
        """Get all stories for a specific JLPT level"""
        return [
            StoryListItem(
                id=story.id,
                title=story.metadata.title,
                titleJapanese=story.metadata.titleJapanese,
                jlptLevel=story.metadata.jlptLevel,
                wordCount=story.metadata.wordCount,
                genre=story.metadata.genre,
                summary=story.metadata.summary,
                coverImageURL=story.metadata.coverImageURL,
                chapterCount=len(story.chapters),
                isPremium=story.metadata.isPremium
            )
            for story in self._stories_cache.values()
            if story.metadata.jlptLevel.upper() == jlpt_level.upper()
        ]

    def reload_stories(self) -> None:
        """Reload all stories from disk"""
        self._stories_cache.clear()
        self._load_stories()


# Global instance
_story_service: Optional[StoryService] = None


def get_story_service() -> StoryService:
    """Get or create the story service instance"""
    global _story_service
    if _story_service is None:
        # Default path - can be configured via environment
        stories_path = Path(__file__).parent.parent / "data" / "stories"
        _story_service = StoryService(stories_path)
    return _story_service
